aten/src/ATen/native/cuda/BatchLinearAlgebra.cu

```diff
@@ -1999,6 +1999,53 @@ Tensor& orgqr_kernel_impl(Tensor& result, const Tensor& tau, int64_t n_columns)
 
 // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ qr ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+template <typename scalar_t>
+static void apply_geqrf(const Tensor& input, const Tensor& tau, int64_t m64, int64_t n64) {
+#ifndef USE_MAGMA
+  TORCH_CHECK(
+    false,
+    "Calling torch.geqrf on a CUDA tensor requires compiling ",
+    "PyTorch with MAGMA. Please use PyTorch built with MAGMA support.");
+#else
+
+  magma_int_t m = magma_int_cast(m64, "m");
+  magma_int_t n = magma_int_cast(n64, "n");
+
+  auto input_data = input.data_ptr<scalar_t>();
+  auto input_matrix_stride = matrixStride(input);
+  auto tau_stride = tau.size(-1);
+  auto batch_size = batchCount(input);
+  auto lda = std::max<int>(1, m);
+
+  // magmaGeqrf uses a hybrid CPU-GPU algorithm to compute the elementary reflectors.
+  // The driver routine geqrf2_gpu accepts a tensor on the CPU for elementary reflectors.
+  Tensor tau_cpu = at::empty(tau.sizes(), tau.options().device(at::kCPU));
```

Comment:
does it make sense to allocate this in pinned memory and use non-blocking copy below? The effect is probably negligible, but I don't know off the top of my head.

Reply:
I think it wouldn't hurt.
Is it simply `Tensor tau_cpu = at::empty(tau.sizes(), tau.options().device(at::kCPU).pinned_memory(true));` ?

Reply:
Yes, and `,/*non_blocking*/true` for `copy_` below, iirc.

Reply:
Thank you! It did give non-negligible speedup:
```
|                              | before | after |
|------------------------------|--------|-------|
| torch.Size([2, 2])           | 3.0    | 2.3   |
| torch.Size([2, 2, 2])        | 5.8    | 4.4   |
| torch.Size([32, 2, 2])       | 88.3   | 70.6  |
| torch.Size([8, 8])           | 2.9    | 2.2   |
| torch.Size([2, 8, 8])        | 5.7    | 4.4   |
| torch.Size([32, 8, 8])       | 88.3   | 70.7  |
| torch.Size([16, 16])         | 3.0    | 2.2   |
| torch.Size([2, 16, 16])      | 5.7    | 4.5   |
| torch.Size([32, 16, 16])     | 89.0   | 72.4  |
| torch.Size([32, 32])         | 3.0    | 2.3   |
| torch.Size([2, 32, 32])      | 5.8    | 4.6   |
| torch.Size([32, 32, 32])     | 91.6   | 72.7  |
| torch.Size([64, 64])         | 3.1    | 2.4   |
| torch.Size([2, 64, 64])      | 6.0    | 4.7   |
| torch.Size([32, 64, 64])     | 93.1   | 74.0  |
| torch.Size([128, 128])       | 3.5    | 2.8   |
| torch.Size([2, 128, 128])    | 6.9    | 5.5   |
| torch.Size([32, 128, 128])   | 107.8  | 88.1  |
| torch.Size([256, 256])       | 4.7    | 3.7   |
| torch.Size([2, 256, 256])    | 9.2    | 7.2   |
| torch.Size([32, 256, 256])   | 144.1  | 112.1 |
| torch.Size([512, 512])       | 6.9    | 6.0   |
| torch.Size([2, 512, 512])    | 13.3   | 11.8  |
| torch.Size([32, 512, 512])   | 209.3  | 184.6 |
| torch.Size([1024, 1024])     | 15.4   | 14.6  |
| torch.Size([2, 1024, 1024])  | 30.5   | 28.9  |
| torch.Size([32, 1024, 1024]) | 486.5  | 462.4 |

Times are in milliseconds (ms).
```

Reply:
I just hope magma is synchronizing when it is writing tau, otherwise nonblocking copy might produce wrong result (it will copy tau to gpu before it's properly populated)